File: resources/solve4.m

% liver data analysis

% load data
A=textread('liver.csv','%s','delimiter',',');          %#ok
A=reshape(A,11,[]);
[r,c]=size(A);

% ceate numerical data arrays
%   data is 10 by ~580 maximum normalized data values
%   class is 1 by ~580 0-1 classifier
data=zeros(r-1,c);
class=zeros(1,c);
idx=[];
for k=1:r
    if k==2
        for j=1:c
            if ~isempty(A{2,j})
                data(k,j)=(length(A{2,j})-4)/2;
            else
                idx=[idx j];                           %#ok
            end
        end
    elseif k==11
        for j=1:c
            try
                class(j)=str2num(A{11,j})-1;           %#ok
            catch 
                idx=[idx j];                           %#ok
            end
        end
    else
        for j=1:c
            try
                data(k,j)=str2num(A{k,j});             %#ok
            catch
                idx=[idx j];                           %#ok
            end
        end
    end
end

% remove instances with missing data
idx=unique(idx);
data(:,idx)=[];
class(idx)=[];

% normalize data to [0,1]
MinValues=min(data,[],2);
MaxValues=max(data,[],2);
data=(data-MinValues)./(MaxValues-MinValues);
[r,c]=size(data);

% subsample features
features=(1:10)';
data=data(features,:);
[r,c]=size(data);

% construct training data 
m=100;
cdx=find(class==1);
ddx=find(class==0);
par.traindata=[data(:,cdx(1:m)) data(:,ddx(1:m))];
sh=1/3;
par.classdata=[class(cdx(1:m)) class(ddx(1:m))]*(1-2*sh)+sh;

% set up NN parameters and initial weights
par.dimensions=[10 10 10 1];
q=length(par.dimensions);
matrixsizes=par.dimensions(1:q-1).*par.dimensions(2:q);
numweights=sum(matrixsizes);
sz=1/sqrt(par.dimensions(1));
w=2*sz*rand(numweights,1)-sz;

% call the optimization
pr.objective=@nnloss;
pr.par=par;
pr.x0=w;
pr.method='BFGS';
pr.linesearch='StrongWolfe';
pr.maxiter=100000;
pr.dftol=1E-9;
pr.ngtol=1E-9;
pr.dxtol=1E-9;
pr.maxcond=1000;
pr.progress=1000;

% This is the call to the optimizer
pr.par.classify=false;
out=optimize(pr);

% This is a call to classify the training data
pr.par.classify=true;
res=pr.objective(out.x(:,end),pr.par);

% Show the confusion matrix for the training data
NN=sum(res<0.5&par.classdata<0.5);
NP=sum(res<0.5&par.classdata>=0.5);
PN=sum(res>=0.5&par.classdata<0.5);
PP=sum(res>=0.5&par.classdata>=0.5);
ConfusionMatrix=[NN PN ; NP PP]                %#ok

% Classify the test data and show the confusion matrix
pr.par.traindata=[data(:,cdx(m+1:end)) data(:,ddx(m+1:end))];
pr.par.classdata=[class(cdx(m+1:end)) class(ddx(m+1:end))]*1/3+1/3;
test=pr.objective(out.x(:,end),pr.par);
NN=sum(test<0.5&pr.par.classdata<0.5);
NP=sum(test<0.5&pr.par.classdata>=0.5);
PN=sum(test>=0.5&pr.par.classdata<0.5);
PP=sum(test>=0.5&pr.par.classdata>=0.5);
ConfusionMatrix=[NN PN ; NP PP]               %#ok

File: resources/solve4.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
script for solving the liver disease project
Version: Aug  14, 2023
Author: Tom Asaki
"""

import numpy as np
import optimize as opt
import pandas as pd              # type: ignore

# read in raw data, drop rows with any missing data
B=pd.read_csv('liver.csv').dropna().to_numpy()
r,c=B.shape
# set 'Male' to 0 and 'Female' to 1
for k in range(r):
    B[k,1]=(len(B[k,1])-4)/2
A=np.zeros((r,c))
for k in range(c):
    A[:,k]=B[:,k]


# data class vector
data_class=A[:,c-1]-1

# normalize data values to be 0 <= val <= 1
data_values=A[:,0:c-1]
maxval=np.array(list(max(data_values[:,k]) for k in range(c-1)))
minval=np.array(list(min(data_values[:,k]) for k in range(c-1)))
data_values=(data_values-minval)/(maxval-minval)

# separate into training and testing data
numtrain=200
idx=np.random.permutation(r)
idx=np.arange(200)
train_data=data_values[idx[0:numtrain],]
train_class=np.array([data_class[idx[0:numtrain]]]).reshape(-1,1)
test_data=data_values[idx[numtrain:r],]
test_class=np.array([data_class[idx[numtrain:r]]]).reshape(-1,1)

# set up optimization problem
#  training data  corresponding class          layer sizes     task
p=[train_data.T , (train_class.T)*(1/3)+(1/3), [10,10,10,1] , 'train']

# set up initial weights
sz=np.sqrt(p[2][0])
NumWeights=0
for k in range(len(p[2])-1):
    NumWeights+=p[2][k]*p[2][k+1]
x0=sz*np.random.randn(NumWeights).reshape((NumWeights,1))

    
from objective import nnloss as obj

alg=dict(obj        = obj,
         x0         = x0,
         params     = p,
         method     = 'LBFGS',
         maxiter    = 20000,
         ngtol      = 1E-10,
         dftol      = 1E-10,
         dxtol      = 1E-10,
         Lambda     = 1,
         Lambdamax  = 100,
         linesearch = 'StrongWolfe',
         c1         = 0.0001,
         c2         = 0.9,
         progress   = 1000
         )   

res=opt.minimize(alg)

alg['params'][3]='classify'
fitclass=obj(res['x'][:,[-1,]],p,1)

comp=np.abs(np.round(fitclass)-train_class.T)
print('Training Misclassification Rate = %3.1f%%' % (np.sum(comp)/np.size(comp)*100))


